src/CustomElements/v1/CustomElements.js

```diff
@@ -0,0 +1,597 @@
+              this._upgradeElement(node, definition, true);
+            }
+            if (node.__upgraded && !node.__attached) {
+              node.__attached = true;
+              if (definition && definition.connectedCallback) {
+                definition.connectedCallback.call(node);
+              }
+            }
+          }
+          if (node.shadowRoot) {
+            // TODO(justinfagnani): do we need to check that the shadowRoot
+            // is observed?
+            this._addNodes(node.shadowRoot.childNodes);
+          }
+          if (node.tagName === 'LINK') {
+            var onLoad = (function() {
+              var link = node;
+              return function() {
+                link.removeEventListener('load', onLoad);
+                this._observeRoot(link.import);
+                this._addNodes(link.import.childNodes);
+              }.bind(this);
+            }).bind(this)();
+            if (node.import) {
+              onLoad();
+            } else {
+              node.addEventListener('load', onLoad);
+            }
+          }
+        } while (walker.nextNode())
+      }
+    },
+
+    /**
+     * @param {NodeList} nodeList
+     * @private
+     */
+    _removeNodes: function(nodeList) {
+      for (var i = 0; i < nodeList.length; i++) {
+        var root = nodeList[i];
+
+        if (!isElement(root)) {
+          continue;
+        }
+
+        // Since we're detatching this element from an observed root, we need to
+        // reobserve it.
+        // TODO(justinfagnani): can we do this in a microtask so we don't thrash
+        // on creating and destroying MutationObservers on batch DOM mutations?
+        this._observeRoot(root);
+
+        var walker = createTreeWalker(root);
+        do {
+          var node = walker.currentNode;
+          if (node.__upgraded && node.__attached) {
+            node.__attached = false;
+            var definition = this._definitions.get(node.localName);
+            if (definition && definition.disconnectedCallback) {
+              definition.disconnectedCallback.call(node);
+            }
+          }
+        } while (walker.nextNode())
+      }
+    },
+
+    /**
+     * Upgrades or customizes a custom element.
+     *
+     * @param {HTMLElement} element
+     * @param {CustomElementDefinition} definition
+     * @param {boolean} callConstructor
+     * @private
+     */
+    _upgradeElement: function(element, definition, callConstructor) {
+      var prototype = definition.constructor.prototype;
+      element.__proto__ = prototype;
+      if (callConstructor) {
+        this._setNewInstance(element);
+        element.__upgraded = true;
+        new (definition.constructor)();
+        console.assert(this._newInstance == null);
+      }
+
+      var observedAttributes = definition.observedAttributes;
+      if (definition.attributeChangedCallback && observedAttributes.length > 0) {
+        this._attributeObserver.observe(element, {
```

Comment:
If we go with the patching strategy, then overriding `setAttribute`/`removeAttribute` is likely good enough.